allennlp/models/ensemble/bidaf_ensemble.py

```diff
@@ -0,0 +1,158 @@
+
+        Returns
+        -------
+        An output dictionary consisting of:
+        best_span : torch.IntTensor
+            The result of a constrained inference over ``span_start_logits`` and
+            ``span_end_logits`` to find the most probable span.  Shape is ``(batch_size, 2)``
+            and each offset is a token index.
+        best_span_str : List[str]
+            If sufficient metadata was provided for the instances in the batch, we also return the
+            string from the original passage that the model thinks is the best answer to the
+            question.
+        """
+
+        subresults = []
+        for submodel in self.submodels:
+            subresults.append(submodel(question, passage, span_start, span_end, metadata))
+
+        batch_size = len(subresults[0]["best_span"])
+
+        output = {
+                "best_span": torch.zeros(batch_size, 2).long()
+        }
+        for index in range(batch_size):
+            best = ensemble(index, subresults)
+            best_span = subresults[best]["best_span"].data[index].long()
+            output["best_span"][index] = best_span
+
+            if metadata is not None:
+                if "best_span_str" not in output:
+                    output["best_span_str"] = []
+                best_span_str = subresults[best]["best_span_str"][index]
+                output["best_span_str"].append(best_span_str)
+
+                answer_texts = metadata[index].get('answer_texts', [])
+                if answer_texts:
+                    self._squad_metrics(best_span_str, answer_texts)
+
+        return output
+
+    def get_metrics(self, reset: bool = False) -> Dict[str, float]:
+        exact_match, f1_score = self._squad_metrics.get_metric(reset)
+        return {
+                'em': exact_match,
+                'f1': f1_score,
+        }
+
+    @classmethod
+    def from_params(cls, vocab: Vocabulary, params: Params):
+        assert not vocab, "vocab should be None"
+
+        submodels = []
+        paths = params.pop("submodels")
+        for path in paths:
+            submodels.append(load_archive(path).model)
+
+        return cls(submodels=submodels)
+
+def ensemble(index, subresults: List[Dict[str, torch.Tensor]]) -> int:
+    """
+    Identifies the best prediction given the results from the submodels.
+
+    Parameters
+    ----------
+    index : the index within this index to ensemble
+    subresults : List[Dict[str, torch.Tensor]]
+
+    Returns
+    -------
+    The index of the best submodel.
+    """
+
+    # Populate span_votes so each key represents a span range that a submodel predicts and the value
+    # is the number of models that made the prediction.
+    spans = [(subresult["best_span"].data[index][0], subresult["best_span"].data[index][1])
+             for subresult in subresults]
+    votes: Dict[Tuple[int, int], int] = {span:spans.count(span) for span in spans}
```

Comment:
you should learn `collections.Counter`

```
votes = collections.Counter(spans)
```

Reply:
I'll read up on it.

Reply:
It doesn't matter much here (because it's unlikely that you'll have more than ~10 sub-models), but counting the spans in this way is n^2 wrt the number of spans. Feel free to ignore. Also, the comment referring to `span_votes` is outdated.